File: ruby/lib/extism.rb
require 'ffi'
require 'json'
require_relative './extism/version'

module Extism
  class Error < StandardError
  end

  # Return the version of Extism
  #
  # @return [String] The version string of the Extism runtime
  def self.extism_version
    C.extism_version
  end

  # Set log file and level, this is a global configuration
  # @param name [String] The path to the logfile
  # @param level [String] The log level. One of {"debug", "error", "info", "trace" }
  def self.set_log_file(name, level = nil)
    C.extism_log_file(name, level)
  end

  $PLUGINS = {}
  $FREE_PLUGIN = proc { |ptr|
    x = $PLUGINS[ptr]
    unless x.nil?
      C.extism_plugin_free(x[:plugin])
      $PLUGINS.delete(ptr)
    end
  }

  # A CancelHandle can be used to cancel a running plugin from another thread
  class CancelHandle
    def initialize(handle)
      @handle = handle
    end

    # Cancel the plugin used to generate the handle
    def cancel
      C.extism_plugin_cancel(@handle)
    end
  end

  # A Plugin represents an instance of your WASM program from the given manifest.
  class Plugin
    # Intialize a plugin
    #
    # @param wasm [Hash, String] The manifest or WASM binary. See https://extism.org/docs/concepts/manifest/.
    # @param wasi [Boolean] Enable WASI support
    # @param config [Hash] The plugin config
    def initialize(wasm, functions = [], wasi = false, config = nil)
      wasm = JSON.generate(wasm) if wasm.instance_of?(Hash)
      code = FFI::MemoryPointer.new(:char, wasm.bytesize)
      errmsg = FFI::MemoryPointer.new(:pointer)
      code.put_bytes(0, wasm)
      funcs_ptr = FFI::MemoryPointer.new(C::ExtismFunction)
      funcs_ptr.write_array_of_pointer(functions.map { |f| f.pointer })
      @plugin = C.extism_plugin_new(code, wasm.bytesize, funcs_ptr, functions.length, wasi, errmsg)
      if @plugin.null?
        err = errmsg.read_pointer.read_string
        C.extism_plugin_new_error_free errmsg.read_pointer
        raise Error, err
      end
      $PLUGINS[object_id] = { plugin: @plugin }
      ObjectSpace.define_finalizer(self, $FREE_PLUGIN)
      return unless !config.nil? and @plugin.null?

      s = JSON.generate(config)
      ptr = FFI::MemoryPointer.from_string(s)
      C.extism_plugin_config(@plugin, ptr, s.bytesize)
    end

    # Check if a function exists
    #
    # @param name [String] The name of the function
    # @return [Boolean] Returns true if function exists
    def has_function?(name)
      C.extism_plugin_function_exists(@plugin, name)
    end

    # Call a function by name
    #
    # @param name [String] The function name
    # @param data [String] The input data for the function
    # @return [String] The output from the function in String form
    def call(name, data, &block)
      # If no block was passed then use Pointer::read_string
      block ||= ->(buf, len) { buf.read_string(len) }
      input = FFI::MemoryPointer.from_string(data)
      rc = C.extism_plugin_call(@plugin, name, input, data.bytesize)
      if rc != 0
        err = C.extism_plugin_error(@plugin)
        raise Error, 'extism_call failed' if err&.empty?

        raise Error, err

      end

      out_len = C.extism_plugin_output_length(@plugin)
      buf = C.extism_plugin_output_data(@plugin)
      block.call(buf, out_len)
    end

    # Free a plugin, this should be called when the plugin is no longer needed
    #
    # @return [void]
    def free
      return if @plugin.null?

      $PLUGINS.delete(object_id)
      C.extism_plugin_free(@plugin)
      @plugin = nil
    end

    # Get a CancelHandle for a plugin
    def cancel_handle
      CancelHandle.new(C.extism_plugin_cancel_handle(@plugin))
    end
  end

  Memory = Struct.new(:offset, :len)

  class CurrentPlugin
    def initialize(ptr)
      @ptr = ptr
    end

    def memory_ptr(mem)
      plugin_ptr = C.extism_current_plugin_memory(@ptr)
      FFI::Pointer.new(plugin_ptr.address + mem.offset)
    end

    def alloc(amount)
      offset = C.extism_current_plugin_memory_alloc(@ptr, amount)
      Memory.new(offset, amount)
    end

    def free(memory)
      C.extism_current_plugin_memory_free(@ptr, memory.offset)
    end

    def memory_at_offset(offset)
      len = C.extism_current_plugin_memory_length(@ptr, offset)
      Memory.new(offset, len)
    end

    def input_as_bytes(input)
      # TODO: should assert that this is an int input
      mem = memory_at_offset(input.value)
      # if mem
      #   require 'debug'
      #   binding.break
      # end
      memory_ptr(mem).read_bytes(mem.len)
    end

    def return_bytes(output, bytes)
      mem = alloc(bytes.length)
      memory_ptr(mem).put_bytes(0, bytes)
      output.value = mem.offset
    end

    def return_string(output, string)
      return_bytes(output, string)
    end
  end

  module ValType
    I32 = 0
    I64 = 1
    F32 = 2
    F64 = 3
    V128 = 4
    FUNC_REF = 5
    EXTERN_REF = 6
  end

  class Val
    def initialize(ptr)
      @c_val = C::ExtismVal.new(ptr)
    end

    def type
      case @c_val[:t]
      when :I32
        :i32
      when :I64
        :i64
      when :F32
        :f32
      when :F64
        :f64
      else
        raise "Unsupported wasm value type #{type}"
      end
    end

    def value
      @c_val[:v][type]
    end

    def value=(val)
      @c_val[:v][type] = val
    end
  end

  class Function
    def initialize(name, args, returns, func_proc, user_data)
      @name = name
      @args = args
      @returns = returns
      @func = func_proc
      @user_data = user_data
    end

    def pointer
      return @pointer if @pointer

      free = proc { puts 'freeing ' }
      args = C.from_int_array(@args)
      returns = C.from_int_array(@returns)
      @pointer = C.extism_function_new(@name, args, @args.length, returns, @returns.length, c_func, free, nil)
    end

    private

    def c_func
      @c_func ||= proc do |plugin_ptr, inputs_ptr, inputs_size, outputs_ptr, outputs_size, _data_ptr|
        current_plugin = CurrentPlugin.new(plugin_ptr)
        val_struct_size = C::ExtismVal.size

        inputs = (0...inputs_size).map do |i|
          Val.new(inputs_ptr + i * val_struct_size)
        end
        outputs = (0...outputs_size).map do |i|
          Val.new(outputs_ptr + i * val_struct_size)
        end

        @func.call(current_plugin, inputs, outputs, @user_data)
      end
    end
  end

  # Private module used to interface with the Extism runtime.
  # *Warning*: Do not use or rely on this directly.
  module C
    extend FFI::Library
    ffi_lib 'extism'

    def self.from_int_array(ruby_array)
      ptr = FFI::MemoryPointer.new(:int, ruby_array.length)
      ptr.write_array_of_int(ruby_array)
      ptr
    end

    typedef :uint64, :ExtismMemoryHandle
    typedef :uint64, :ExtismSize

    enum :ExtismValType, %i[I32 I64 F32 F64 V128 FuncRef ExternRef]

    class ExtismValUnion < FFI::Union
      layout :i32, :int32,
             :i64, :int64,
             :f32, :float,
             :f64, :double
    end

    class ExtismVal < FFI::Struct
      layout :t, :ExtismValType,
             :v, ExtismValUnion
    end

    class ExtismFunction < FFI::Struct
      layout :name, :string,
             :inputs, :pointer,
             :n_inputs, :uint64,
             :outputs, :pointer,
             :n_outputs, :uint64,
             :data, :pointer
    end

    callback :ExtismFunctionType, [
      :pointer, # plugin
      :pointer, # inputs
      :ExtismSize, # n_inputs
      :pointer, # outputs
      :ExtismSize, # n_outputs
      :pointer # user_data
    ], :void

    callback :ExtismFreeFunctionType, [], :void

    attach_function :extism_plugin_id, [:pointer], :pointer
    attach_function :extism_current_plugin_memory, [:pointer], :pointer
    attach_function :extism_current_plugin_memory_alloc, %i[pointer ExtismSize], :ExtismMemoryHandle
    attach_function :extism_current_plugin_memory_length, %i[pointer ExtismMemoryHandle], :ExtismSize
    attach_function :extism_current_plugin_memory_free, %i[pointer ExtismMemoryHandle], :void
    attach_function :extism_function_new,
                    %i[string pointer ExtismSize pointer ExtismSize ExtismFunctionType ExtismFreeFunctionType pointer], :pointer
    attach_function :extism_function_free, [:pointer], :void
    attach_function :extism_function_set_namespace, %i[pointer string], :void
    attach_function :extism_plugin_new, %i[pointer ExtismSize pointer ExtismSize bool pointer], :pointer
    attach_function :extism_plugin_new_error_free, [:pointer], :void
    attach_function :extism_plugin_free, [:pointer], :void
    attach_function :extism_plugin_cancel_handle, [:pointer], :pointer
    attach_function :extism_plugin_cancel, [:pointer], :bool
    attach_function :extism_plugin_config, %i[pointer pointer ExtismSize], :bool
    attach_function :extism_plugin_function_exists, %i[pointer string], :bool
    attach_function :extism_plugin_call, %i[pointer string pointer ExtismSize], :int32
    attach_function :extism_error, [:pointer], :string
    attach_function :extism_plugin_error, [:pointer], :string
    attach_function :extism_plugin_output_length, [:pointer], :ExtismSize
    attach_function :extism_plugin_output_data, [:pointer], :pointer
    attach_function :extism_log_file, %i[string string], :bool
    attach_function :extism_version, [], :string
  end
end

File: haskell/src/Extism.hs
module Extism (
  module Extism.Manifest,
  ValType(..),
  Val(..),
  Function(..),
  Plugin(..),
  CancelHandle(..),
  CurrentPlugin(..),
  LogLevel(..),
  Error(..),
  Result(..),
  toByteString,
  fromByteString,
  extismVersion,
  plugin,
  pluginFromManifest,
  isValid,
  setConfig,
  setLogFile,
  functionExists,
  hostFunction,
  call,
  cancelHandle,
  cancel,
  pluginID,
  unwrap
) where

import Foreign.ForeignPtr
import Foreign.C.String
import Foreign.Ptr
import Foreign.Marshal.Array
import Foreign.Marshal.Alloc
import Foreign.Storable
import Foreign.StablePtr
import Foreign.Concurrent
import qualified Data.ByteString as B
import qualified Data.ByteString.Lazy as BL
import Data.ByteString.Internal (c2w, w2c)
import Data.ByteString.Unsafe (unsafeUseAsCString)
import qualified Text.JSON (encode, toJSObject, showJSON)
import Extism.Manifest (Manifest, toString)
import Extism.Bindings
import qualified Data.UUID (UUID, fromByteString)

-- | Host function, see 'Extism.HostFunction.hostFunction'
data Function = Function (ForeignPtr ExtismFunction) (StablePtr ())

-- | Plugins can be used to call WASM function
newtype Plugin = Plugin (ForeignPtr ExtismPlugin)

-- | Cancellation handle for Plugins
newtype CancelHandle = CancelHandle (Ptr ExtismCancelHandle)

-- | Access the plugin that is currently executing from inside a host function
type CurrentPlugin = Ptr ExtismCurrentPlugin

-- | Log level
data LogLevel = LogError | LogWarn | LogInfo | LogDebug | LogTrace deriving (Show)

-- | Extism error
newtype Error = ExtismError String deriving Show

-- | Result type
type Result a = Either Error a

-- | Helper function to convert a 'String' to a 'ByteString'
toByteString :: String -> B.ByteString
toByteString x = B.pack (map c2w x)

-- | Helper function to convert a 'ByteString' to a 'String'
fromByteString :: B.ByteString -> String
fromByteString bs = map w2c $ B.unpack bs

-- | Get the Extism version string
extismVersion :: () -> IO String
extismVersion () = do
  v <- extism_version
  peekCString v

-- | Create a 'Plugin' from a WASM module, `useWasi` determines if WASI should
-- | be linked
plugin :: B.ByteString -> [Function] -> Bool -> IO (Result Plugin)
plugin wasm functions useWasi =
  let nfunctions = fromIntegral (length functions) in
  let length = fromIntegral (B.length wasm) in
  let wasi = fromInteger (if useWasi then 1 else 0) in
  do
    funcs <- mapM (\(Function ptr _) -> withForeignPtr ptr (\x -> do return x)) functions
    alloca (\e-> do
      let errmsg = (e :: Ptr CString)
      p <- unsafeUseAsCString wasm (\s ->
        withArray funcs (\funcs ->
          extism_plugin_new (castPtr s) length funcs nfunctions wasi errmsg ))
      if p == nullPtr then do
        err <- peek errmsg
        e <- peekCString err 
        extism_plugin_new_error_free err
        return $ Left (ExtismError e)
      else do
        ptr <- Foreign.Concurrent.newForeignPtr p (extism_plugin_free p)
        return $ Right (Plugin ptr))

-- | Create a 'Plugin' from a 'Manifest'
pluginFromManifest :: Manifest -> [Function] -> Bool -> IO (Result Plugin)
pluginFromManifest manifest functions useWasi =
  let wasm = toByteString $ toString manifest in
  plugin wasm functions useWasi

-- | Check if a 'Plugin' is valid
isValid :: Plugin -> IO Bool
isValid (Plugin p) = withForeignPtr p (\x -> return (x /= nullPtr))

-- | Set configuration values for a plugin
setConfig :: Plugin -> [(String, Maybe String)] -> IO Bool
setConfig (Plugin plugin) x =
  let obj = Text.JSON.toJSObject [(k, Text.JSON.showJSON v) | (k, v) <- x] in
  let bs = toByteString (Text.JSON.encode obj) in
  let length = fromIntegral (B.length bs) in
  unsafeUseAsCString bs (\s -> do
    withForeignPtr plugin (\plugin-> do
      b <- extism_plugin_config plugin (castPtr s) length
      return $ b /= 0))

levelStr LogError = "error"
levelStr LogDebug = "debug"
levelStr LogWarn = "warn"
levelStr LogTrace = "trace"
levelStr LogInfo = "info"

-- | Set the log file and level, this is a global configuration
setLogFile :: String -> LogLevel -> IO Bool
setLogFile filename level =
  let s = levelStr level in
  withCString filename (\f ->
    withCString s (\l -> do
      b <- extism_log_file f l
      return $ b /= 0))

-- | Check if a function exists in the given plugin
functionExists :: Plugin -> String -> IO Bool
functionExists (Plugin plugin) name = do
  withForeignPtr plugin (\plugin -> do
    b <- withCString name (extism_plugin_function_exists plugin)
    if b == 1 then return True else return False)

--- | Call a function provided by the given plugin
call :: Plugin -> String -> B.ByteString -> IO (Result B.ByteString)
call (Plugin plugin) name input =
  let length = fromIntegral (B.length input) in
  do
    withForeignPtr plugin (\plugin -> do
      rc <- withCString name (\name ->
        unsafeUseAsCString input (\input ->
          extism_plugin_call plugin name (castPtr input) length))
      err <- extism_error plugin
      if err /= nullPtr
        then do e <- peekCString err
                return $ Left (ExtismError e)
      else if rc == 0
        then do
          length <- extism_plugin_output_length plugin
          ptr <- extism_plugin_output_data plugin
          buf <- B.packCStringLen (castPtr ptr, fromIntegral length)
          return $ Right buf
      else return $ Left (ExtismError "Call failed"))

-- | Call a function with a string argument and return a string
callString :: Plugin -> String -> String -> IO (Result String)
callString p name input = do
  res <- call p name (toByteString input)
  case res of 
    Left x -> return $ Left x
    Right x -> return $ Right (fromByteString x)


-- | Create a new 'CancelHandle' that can be used to cancel a running plugin
-- | from another thread.
cancelHandle :: Plugin -> IO CancelHandle
cancelHandle (Plugin plugin) = do
  handle <- withForeignPtr plugin extism_plugin_cancel_handle
  return (CancelHandle handle)

-- | Cancel a running plugin using a 'CancelHandle'
cancel :: CancelHandle -> IO Bool
cancel (CancelHandle handle) =
  extism_plugin_cancel handle


-- | Create a new 'Function' that can be called from a 'Plugin'
hostFunction :: String -> [ValType] -> [ValType] -> (CurrentPlugin -> [Val] -> a -> IO [Val]) -> a -> IO Function
hostFunction name params results f v =
  let nparams = fromIntegral $ length params in
  let nresults = fromIntegral $ length results in
  do
    cb <- callbackWrap (callback f :: CCallback)
    free <- freePtrWrap freePtr
    userData <- newStablePtr (v, free, cb)
    let userDataPtr = castStablePtrToPtr userData
    x <- withCString name (\name ->  do
      withArray params (\params ->
        withArray results (\results -> do
          extism_function_new name params nparams results nresults cb userDataPtr free)))
    let freeFn = extism_function_free x
    fptr <- Foreign.Concurrent.newForeignPtr x freeFn
    return $ Function fptr (castPtrToStablePtr userDataPtr)


pluginID :: Plugin -> IO Data.UUID.UUID
pluginID (Plugin plugin) =
  withForeignPtr plugin (\plugin -> do
    ptr <- extism_plugin_id plugin
    buf <- B.packCStringLen (castPtr ptr, 16)
    case Data.UUID.fromByteString (BL.fromStrict buf) of
      Nothing -> error "Invalid Plugin ID" 
      Just x -> return x)

    
unwrap (Right x) = x
unwrap (Left (ExtismError msg)) = do
  error msg
